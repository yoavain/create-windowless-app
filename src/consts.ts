export const consts = {
    dependencies: [
        "node-notifier",
        "winston"
    ],
    devDependencies: [
        "fs-extra",
        "jest",
        "webpack",
        "webpack-cli",
        "copy-webpack-plugin",
        "rimraf",
        "cross-spawn"
    ],
    huskyDependencies: [
        "husky"
    ],
    tsDevDependencies: [
        "@types/jest",
        "@types/node",
        "@tsconfig/node18",
        "@types/node-notifier",
        "@types/winston",
        "ts-loader",
        "ts-node",
        "typescript",
        "@types/cross-spawn"
    ],
    errorLogFilePatterns: [
        "npm-debug.log"
    ],
    validFiles: [
        ".DS_Store",
        "Thumbs.db",
        ".git",
        ".gitignore",
        ".idea",
        "README.md",
        "LICENSE",
        ".hg",
        ".hgignore",
        ".hgcheck",
        ".npmignore",
        "mkdocs.yml",
        "docs",
        ".travis.yml",
        ".gitlab-ci.yml",
        ".gitattributes"
    ],
    knownGeneratedFiles: [
        "package.json",
        "webpack.config.js",
        "tsconfig.json",
        "tsconfig.build.json",
        "src",
        "resources",
        "launcher",
        "node_modules"
    ]
};
