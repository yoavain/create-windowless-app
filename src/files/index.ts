export * from "./fileManager";
export * from "./fileUtils";
