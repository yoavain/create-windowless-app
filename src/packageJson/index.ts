export * from "./packageJsonBuilder";
export * from "./packageJsonConsts";