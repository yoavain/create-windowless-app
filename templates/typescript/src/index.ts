import * as path from "path";
import * as winston from "winston";
import { WindowsToaster } from "node-notifier";
import { execFile } from "child_process";
import type { Option } from "node-notifier";

// App Name
const AppName: string = "##APPNAME##";

const executable: string = process.argv[0];

// Args (ignore exe + js)
const argv: string[] = process.argv.slice(2);

// Logger init
const { combine, timestamp, printf, label } = winston.format;
const filename: string = `${AppName}.log`;
const transports = {
    file: new winston.transports.File({ filename: filename })
};
transports.file.level = "debug";
const logger: winston.Logger = winston.createLogger({
    level: "debug",
    format: combine(
        label({ label: "[my-label]" }),
        timestamp({ format: "YYYY-MM-DD HH:mm:ss.SSS" }),
        printf((info) => `${info.timestamp} [${info.level.toUpperCase()}] ${info.message}`)
    ),
    transports: [transports.file]
});

// Notifier init
const snoreToastPath: string = executable.endsWith(".exe") ? path.resolve(executable, "../", "snoretoast-x64.exe") : null;
let notifierOptions: Option = { withFallback: false, customPath: snoreToastPath };
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const notifier: any = new WindowsToaster(notifierOptions);

// Log message
logger.log("info", `"${AppName}" started with ${argv ? argv.join("; ") : "no args"}`);
logger.log("info", `Notifier started with options ${JSON.stringify(notifierOptions)}`);

// Notify
const notification = { title: `${AppName}`, message: "Hello World", actions: ["Log", "Close"] };
notifier.notify(notification);
notifier.on("log", () => {
    const file: string = path.join(__dirname, "..", filename);
    execFile(file, { shell: "powershell" });
});
