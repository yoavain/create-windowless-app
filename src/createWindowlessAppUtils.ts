import type { InvalidNames, LegacyNames } from "validate-npm-package-name";
import validateProjectName from "validate-npm-package-name";
import chalk from "chalk";
import path from "path";
import { consts } from "./consts";
import { readdirSync, removeSync } from "fs-extra";

// These files should be allowed to remain on a failed install, but then silently removed during the next create.
const errorLogFilePatterns = consts.errorLogFilePatterns;

export const PACKAGE_JSON_FILENAME = "package.json";

// If project only contains files generated by GH, itâ€™s safe.
// Also, if project contains remnant error logs from a previous installation, lets remove them now.
// We also special case IJ-based products .idea because it integrates with CRA:
// https://github.com/facebook/create-react-app/pull/368#issuecomment-243446094
export const isSafeToCreateProjectIn = (root: string, name: string): boolean => {
    const validFiles: string[] = consts.validFiles;
    console.log();

    const conflicts = readdirSync(root)
        .filter((file) => !validFiles.includes(file))
        // IntelliJ IDEA creates module files before CRA is launched
        .filter((file) => !/\.iml$/.test(file))
        // Don't treat log files from previous installation as conflicts
        .filter((file) => !errorLogFilePatterns.some((pattern) => file.indexOf(pattern) === 0));

    if (conflicts.length > 0) {
        console.log(`The directory ${chalk.green(name)} contains files that could conflict:`);
        console.log();
        for (const file of conflicts) {
            console.log(`  ${file}`);
        }
        console.log();
        console.log("Either try using a new directory name, or remove the files listed above.");

        return false;
    }

    // Remove any remnant files from a previous installation
    const currentFiles = readdirSync(path.join(root));
    currentFiles.forEach((file) => {
        errorLogFilePatterns.forEach((errorLogFilePattern) => {
            // This will catch `npm-debug.log*` files
            if (file.indexOf(errorLogFilePattern) === 0) {
                removeSync(path.join(root, file));
            }
        });
    });
    return true;
};

const printValidationResults = (results) => {
    if (typeof results !== "undefined") {
        results.forEach((error) => {
            console.error(chalk.red(`  *  ${error}`));
        });
    }
};

export const checkAppName = (appName) => {
    const validationResult = validateProjectName(appName);
    if (!validationResult.validForNewPackages) {
        console.error(`Could not create a project called ${chalk.red(`"${appName}"`)} because of npm naming restrictions:`);
        printValidationResults((validationResult as InvalidNames).errors);
        printValidationResults((validationResult as LegacyNames).warnings);
        process.exit(1);
    }

    const dependencies = [...consts.dependencies, ...consts.devDependencies].sort();
    if (dependencies.indexOf(appName) >= 0) {
        console.error(
            chalk.red(
                `We cannot create a project called ${chalk.green(appName)} because a dependency with the same name exists.\n` + "Due to the way npm works, the following names are not allowed:\n\n"
            ) +
            chalk.cyan(dependencies.map((depName) => `  ${depName}`).join("\n")) +
            chalk.red("\n\nPlease choose a different project name.")
        );
        process.exit(1);
    }
};

export const replaceAppNamePlaceholder = (appName: string, str: string): string => {
    return str.replace(/##APPNAME##/g, `${appName}`);
};
