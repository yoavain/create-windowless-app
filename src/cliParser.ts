import chalk from "chalk";
import { PACKAGE_JSON_FILENAME } from "./createWindowlessAppUtils";
import { pathExistsSync } from "fs-extra";
import yargs from "yargs";
import { hideBin } from "yargs/helpers";
import { interactiveMode } from "./interactive";
import { validateProjectNameInput } from "./validation";

const packageJson = require(`../${PACKAGE_JSON_FILENAME}`);

export type ProgramConfig = {
    projectName: string;
    icon?: string;
    typescript: boolean;
    husky: boolean;
    skipInstall: boolean;
    nodeVersion: string;
    verbose: boolean;
};

const validateInput = (argv): any => {
    if (argv.icon && !pathExistsSync(argv.icon)) {
        console.log(`Cannot find icon in ${chalk.red(argv.icon)}. Switching to ${chalk.green("default")} icon.`);
        delete argv.icon;
    }

    return argv;
};

export const parseCommand = async (argv: string[]): Promise<ProgramConfig> => {
    const command: any = yargs(hideBin(argv))
        .command("* [projectName]", "project name",
            (yargs) => {
                return yargs.positional("projectName", {
                    describe: "project name",
                    type: "string"
                });
            },
            () => {})
        .option("verbose", {
            alias: "v",
            type: "boolean",
            description: "print additional logs"
        })
        .option("interactive", {
            type: "boolean",
            alias: "i",
            description: "interactive mode"
        })
        .option("typescript", {
            alias: "t",
            type: "boolean",
            description: "use typescript",
            default: true
        })
        .option("husky", {
            alias: "h",
            type: "boolean",
            description: "install husky pre-commit hook for building launcher",
            default: true
        })
        .option("skip-install", {
            alias: "s",
            type: "boolean",
            description: "write dependencies to package.json without installing"
        })
        .option("icon", {
            alias: "c",
            type: "string",
            description: "override default launcher icon file"
        })
        .option("node-version", {
            alias: "n",
            type: "string",
            description: "override node version to bundle"
        })
        .check(({ projectName, interactive, help }) => {
            if (projectName && typeof validateProjectNameInput(projectName) === "string") {
                throw new Error("Invalid project name");
            }
            else if (!projectName && !interactive && !help) {
                throw new Error("Missing project name");
            }
            return true;
        })
        .version("version", packageJson.version)
        .help()
        .middleware(validateInput)
        .strict()
        .argv;

    let programConfig: ProgramConfig;
    if (command.interactive) {
        programConfig = await interactiveMode();
    }
    else {
        programConfig = {
            projectName : command.projectName,
            verbose: command.verbose,
            typescript: command.typescript,
            husky: command.husky,
            skipInstall: command["skip-install"],
            icon: command.icon,
            nodeVersion: command["node-version"]
        };
    }

    return programConfig;
};
