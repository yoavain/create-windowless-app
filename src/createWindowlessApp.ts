import chalk from "chalk";
import * as path from "path";
import { ensureDirSync, readdirSync, removeSync } from "fs-extra";
import spawn from "cross-spawn";
import { compileLauncher } from "./launcherCompiler";
import { consts } from "./consts";
import { checkAppName, getSingleExecutableApplicationsScripts, isSafeToCreateProjectIn, mergeIntoPackageJson, replaceAppNamePlaceholder } from "./createWindowlessAppUtils";
import { copyFile, readJsonResource, readResource, writeFile, writeJson } from "./fileUtils";
import { checkThatNpmCanReadCwd } from "./nodeUtils";
import type { ProgramConfig } from "./cliParser";
import { parseCommand } from "./cliParser";

// TypeScript
const tsWebpackConfigResourceLocation = "../templates/typescript/webpack.config.ts";
const tsConfigBuildResourceLocation = "../templates/typescript/tsconfig.build.json";
const tsConfigResourceLocation = "../templates/typescript/tsconfig.json";
const tsIndexResourceLocation = "../templates/typescript/src/index.ts";
const tsLauncherCompilerLocation = "../templates/typescript/launcher/launcherCompiler.ts";
const tsCheckNodeVersionLocation = "../templates/typescript/utils/checkNodeVersion.ts";

// JavaScript
const jsWebpackConfigResourceLocation = "../templates/javascript/webpack.config.js";
const jsIndexResourceLocation = "../templates/javascript/src/index.js";
const jsLauncherCompilerLocation = "../templates/javascript/launcher/launcherCompiler.js";
const jsCheckNodeVersionLocation = "../templates/javascript/utils/checkNodeVersion.js";

// Launcher Source
const launcherSrcResourceLocation = "../templates/common/src/launcher.cs";
const launcherSrcModifiedLocation = "launcher/launcher.cs";
const launcherProjResourceLocation = "../templates/common/launcher.csproj";
const launcherProjModifiedLocation = "launcher/launcher.csproj";

// Single executable applications config
const seaConfigResourceLocation = "../templates/common/sea-config.json";

// Default icon location
const defaultLauncherIconLocation = "../templates/common/resources/windows-launcher.ico";

// Husky config file
const huskyConfigFileLocation = "../templates/common/husky/pre-commit";


const install = async (root: string, dependencies: string[], isDev: boolean, programConfig: ProgramConfig): Promise<void> => {
    const { verbose, skipInstall } = programConfig;

    if (!skipInstall) {
        const command = "npm";
        const args = ["install", isDev ? "--save-dev" : "--save", "--save-exact", "--loglevel", "error"].concat(dependencies);
        if (verbose) {
            args.push("--verbose");
        }
        console.log(`Installing ${chalk.green(isDev ? "dev dependencies" : "dependencies")}.`);
        console.log();

        spawn.sync(command, args, { stdio: "inherit" });
    }
    else {
        console.log(`Adding ${chalk.green(isDev ? "dev dependencies" : "dependencies")} to package.json (skipping installation)`);
        console.log();

        const dependenciesObject = dependencies.reduce<Record<string, string>>((acc, dep) => {
            let depName: string = dep;
            let depVersion: string = "^x.x.x";
            if (dep.lastIndexOf("@") > 0) {
                depName = dep.substring(0, dep.lastIndexOf("@"));
                const depVersionString: string[] = dep.substring(dep.lastIndexOf("@")).split(".");
                depVersion = `^${depVersionString?.[0] ?? "x"}.${depVersionString?.[1] ?? "x"}.${depVersionString?.[2] ?? "x"}`;
            }
            acc[depName] = depVersion;
            return acc;
        }, {});
        mergeIntoPackageJson(root, isDev ? "devDependencies" : "dependencies", dependenciesObject);
    }
};

const buildTypeScriptProject = (root: string, appName: string, husky: boolean): void => {
    console.log(`Building project ${chalk.green("files")}.`);
    console.log();

    writeJson(path.resolve(root, "tsconfig.build.json"), readJsonResource(tsConfigBuildResourceLocation));
    writeJson(path.resolve(root, "tsconfig.json"), readJsonResource(tsConfigResourceLocation));
    writeFile(path.resolve(root, "webpack.config.ts"), replaceAppNamePlaceholder(readResource(tsWebpackConfigResourceLocation), appName));
    ensureDirSync(path.resolve(root, "src"));
    writeFile(path.resolve(root, "src", "index.ts"), replaceAppNamePlaceholder(readResource(tsIndexResourceLocation), appName));
    ensureDirSync(path.resolve(root, "utils"));
    writeFile(path.resolve(root, "utils", "checkNodeVersion.ts"), readResource(tsCheckNodeVersionLocation));

    // Add scripts
    const scripts: Record<string, string> = {
        "start": "ts-node src/index.ts",
        "type-check": "tsc --build tsconfig.json",
        "prewebpack": "rimraf build && rimraf dist",
        "webpack": "webpack",
        "prebuild": "npm run check-node-version",
        "build": "npm run type-check && npm run webpack && npm run node-sea",
        "check-node-version": "ts-node -e \"require(\"\"./utils/checkNodeVersion\"\").checkNodeRuntimeVersion()\"",
        "check-msbuild": "ts-node -e \"require(\"\"./launcher/launcherCompiler\"\").checkMsbuildInPath(true)\"",
        "rebuild-launcher": "msbuild launcher/launcher.csproj",
        ...getSingleExecutableApplicationsScripts(appName)
    };

    // Add husky
    if (husky) {
        scripts["prepare"] = "git config --get core.hookspath || husky install";
        scripts["pre-commit"] = `git diff HEAD --exit-code --stat launcher/* || npm run check-msbuild && npm run rebuild-launcher && git add resources/bin/${appName}-launcher.exe`;

        ensureDirSync(path.resolve(root, ".husky"));
        writeFile(path.resolve(root, ".husky", "pre-commit"), readResource(huskyConfigFileLocation));
    }

    mergeIntoPackageJson(root, "scripts", scripts);
};

const buildJavaScriptProject = (root: string, appName: string, husky: boolean): void => {
    console.log(`Building project ${chalk.green("files")}.`);
    console.log();

    writeFile(path.resolve(root, "webpack.config.js"), replaceAppNamePlaceholder(readResource(jsWebpackConfigResourceLocation), appName));
    ensureDirSync(path.resolve(root, "src"));
    writeFile(path.resolve(root, "src", "index.js"), replaceAppNamePlaceholder(readResource(jsIndexResourceLocation), appName));
    ensureDirSync(path.resolve(root, "utils"));
    writeFile(path.resolve(root, "utils", "checkNodeVersion.js"), readResource(jsCheckNodeVersionLocation));

    // Add scripts
    const scripts: Record<string, string> = {
        "start": "node src/index.js",
        "prewebpack": "rimraf build && rimraf dist",
        "webpack": "webpack",
        "prebuild": "npm run check-node-version",
        "build": "npm run webpack && npm run node-sea",
        "check-node-version": "node -e \"require(\"\"./utils/checkNodeVersion\"\").checkNodeRuntimeVersion()\"",
        "check-msbuild": "node -e \"require(\"\"./launcher/launcherCompiler\"\").checkMsbuildInPath(true)\"",
        "rebuild-launcher": "msbuild launcher/launcher.csproj",
        ...getSingleExecutableApplicationsScripts(appName)
    };

    // Add husky
    if (husky) {
        scripts["prepare"] = "git config --get core.hookspath || husky install";
        scripts["pre-commit"] = `git diff HEAD --exit-code --stat launcher/* || npm run check-msbuild && npm run rebuild-launcher && git add resources/bin/${appName}-launcher.exe`;

        ensureDirSync(path.resolve(root, ".husky"));
        writeFile(path.resolve(root, ".husky", "pre-commit"), readResource(huskyConfigFileLocation));
    }

    mergeIntoPackageJson(root, "scripts", scripts);
};

export const buildLauncher = (root: string, appName: string, icon: string, typescript: boolean): Promise<void> => {
    console.log(`Building project ${chalk.green("launcher")}.`);
    console.log();

    ensureDirSync(path.resolve("launcher"));
    writeFile(path.resolve(launcherSrcModifiedLocation), replaceAppNamePlaceholder(readResource(launcherSrcResourceLocation), appName));
    writeFile(path.resolve(launcherProjModifiedLocation), replaceAppNamePlaceholder(readResource(launcherProjResourceLocation), appName));
    if (typescript) {
        copyFile(path.resolve(__dirname, tsLauncherCompilerLocation), path.resolve(root, "launcher", "launcherCompiler.ts"));
    }
    else {
        copyFile(path.resolve(__dirname, jsLauncherCompilerLocation), path.resolve(root, "launcher", "launcherCompiler.js"));
    }

    // Resolve icon
    let iconLocation: string;
    if (icon) {
        iconLocation = path.resolve(icon);
        console.log(`Building launcher with icon: ${chalk.green(icon)}.`);
    }
    else {
        iconLocation = path.resolve(__dirname, defaultLauncherIconLocation);
        console.log(`Building launcher with ${chalk.green("default")} icon.`);
    }
    copyFile(iconLocation, path.resolve(root, "launcher", "launcher.ico"));

    return compileLauncher();
};

const run = async (root: string, appName: string, originalDirectory: string, programConfig: ProgramConfig): Promise<void> => {
    const { typescript, husky, icon } = programConfig;
    const dependencies = [...consts.dependencies];
    const devDependencies = [...consts.devDependencies];
    if (typescript) {
        devDependencies.push(...consts.tsDevDependencies);
    }
    if (husky) {
        devDependencies.push(...consts.huskyDependencies);
    }

    try {
        await install(root, dependencies, false, programConfig);
        await install(root, devDependencies, true, programConfig);
        if (typescript) {
            buildTypeScriptProject(root, appName, husky);
        }
        else {
            buildJavaScriptProject(root, appName, husky);
        }

        // Launcher
        ensureDirSync(path.resolve(root, "resources", "bin"));
        await buildLauncher(root, appName, icon, typescript);

        console.log("Done");
    }
    catch (reason) {
        console.log();
        console.log("Aborting installation.");
        if (reason.command) {
            console.log(`  ${chalk.cyan(reason.command)} has failed.`);
        }
        else {
            console.log(chalk.red("Unexpected error. Please report it as a bug:"));
            console.log(reason);
        }
        console.log();

        // On 'exit' we will delete these files from target directory.
        const knownGeneratedFiles = [...consts.knownGeneratedFiles];
        const currentFiles = readdirSync(path.join(root));
        currentFiles.forEach((file) => {
            knownGeneratedFiles.forEach((fileToMatch) => {
                // This removes all knownGeneratedFiles.
                if (file === fileToMatch) {
                    console.log(`Deleting generated file... ${chalk.cyan(file)}`);
                    removeSync(path.join(root, file));
                }
            });
        });
        const remainingFiles = readdirSync(path.join(root));
        if (!remainingFiles.length) {
            // Delete target folder if empty
            console.log(`Deleting ${chalk.cyan(`${appName}/`)} from ${chalk.cyan(path.resolve(root, ".."))}`);
            process.chdir(path.resolve(root, ".."));
            removeSync(path.join(root));
        }
        console.log("Done (with errors).");
        process.exit(1);
    }
};

const createApp = (programConfig: ProgramConfig): Promise<void> => {
    const { projectName } = programConfig;
    const root: string = path.resolve(projectName);
    const appName: string = path.basename(root);

    checkAppName(appName);
    ensureDirSync(projectName);
    if (!isSafeToCreateProjectIn(root, projectName)) {
        process.exit(1);
    }

    console.log(`Creating a new windowless app in ${chalk.green(root)}.`);
    console.log();

    const packageJson = {
        name: appName,
        version: "0.1.0",
        private: true,
        main: "_build/index.js"
    };
    writeJson(path.join(root, "package.json"), packageJson);
    copyFile(path.resolve(__dirname, seaConfigResourceLocation), path.resolve(root, "sea-config.json"));

    const originalDirectory = process.cwd();
    process.chdir(root);
    if (!checkThatNpmCanReadCwd()) {
        process.exit(1);
    }

    return run(root, appName, originalDirectory, programConfig);
};

export const createWindowlessApp = async (argv: string[]): Promise<void> => {
    const programConfig: ProgramConfig = await parseCommand(argv);

    if (programConfig.projectName) {
        return createApp(programConfig);
    }
};
