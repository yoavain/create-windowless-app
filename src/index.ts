#!/usr/bin/env node
import { createWindowlessApp } from "./createWindowlessApp";
import { checkCscInPath } from "./launcherCompiler";

const currentNodeVersion: string = process.versions.node;
const semver: string[] = currentNodeVersion.split(".");
const major = Number(semver[0]);

if (isNaN(major) || major < 10) {
    console.error(`You are running Node ${currentNodeVersion}.\nCreate Windowless App requires Node 8 or higher.\nPlease update your version of Node.`);
    process.exit(1);
}

// Check for csc.exe in %PATH%
checkCscInPath(true).then(() => {
    // noinspection JSIgnoredPromiseFromCall
    createWindowlessApp(process.argv);
});
