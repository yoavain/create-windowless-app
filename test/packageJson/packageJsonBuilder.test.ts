import { PackageJsonBuilder } from "../../src/packageJson";

describe("Test PackageJsonBuilder", () => {
    it("Test default package.json", () => {
        const builder = new PackageJsonBuilder("test-app");
        expect(builder.build()).toMatchSnapshot();
    });
    it("Test package.json with javascript", () => {
        const builder = new PackageJsonBuilder("test-app").withJavaScript();
        expect(builder.build()).toMatchSnapshot();
    });
    it("Test package.json with husky", () => {
        const builder = new PackageJsonBuilder("test-app").withHusky();
        expect(builder.build()).toMatchSnapshot();
    });
    it("Test package.json with javascript after husky", () => {
        const builder = new PackageJsonBuilder("test-app").withHusky().withJavaScript();
        expect(builder.build()).toMatchSnapshot();
    });
    it("Test package.json with husky after javascript", () => {
        const builder = new PackageJsonBuilder("test-app").withJavaScript().withHusky();
        expect(builder.build()).toMatchSnapshot();
    });
});
