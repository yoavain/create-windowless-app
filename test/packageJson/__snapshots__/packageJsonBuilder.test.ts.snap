// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test PackageJsonBuilder Test default package.json 1`] = `
{
  "main": "_build/index.js",
  "name": "test-app",
  "private": true,
  "scripts": {
    "build": "npm run type-check && npm run webpack && npm run node-sea",
    "check-msbuild": "ts-node -e "require(""./launcher/launcherCompiler"").checkMsbuildInPath(true)"",
    "check-node-version": "ts-node -e "require(""./utils/checkNodeVersion"").checkNodeRuntimeVersion()"",
    "node-sea": "npm run node-sea:build-blob && npm run node-sea:copy-node && npm run node-sea:unsign && npm run node-sea:inject-blob",
    "node-sea:build-blob": "node --experimental-sea-config sea-config.json",
    "node-sea:copy-node": "node -e "require('fs').copyFileSync(process.execPath, 'dist/test-app.exe')"",
    "node-sea:inject-blob": "postject dist/test-app.exe NODE_SEA_BLOB _blob\\sea-prep.blob --sentinel-fuse NODE_SEA_FUSE_fce680ab2cc467b6e072b8b5df1996b2",
    "node-sea:sign": "signtool sign /fd SHA256 dist/test-app.exe",
    "node-sea:unsign": "signtool remove /s dist/test-app.exe || echo Warning: signtool not found in "Path"",
    "prebuild": "npm run check-node-version",
    "prenode-sea:build-blob": "rimraf _blob && mkdir _blob",
    "prewebpack": "rimraf build && rimraf dist",
    "rebuild-launcher": "msbuild launcher/launcher.csproj",
    "start": "ts-node src/index.ts",
    "type-check": "tsc --build tsconfig.json",
    "webpack": "webpack",
  },
  "version": "0.1.0",
}
`;

exports[`Test PackageJsonBuilder Test package.json with husky 1`] = `
{
  "main": "_build/index.js",
  "name": "test-app",
  "private": true,
  "scripts": {
    "build": "npm run type-check && npm run webpack && npm run node-sea",
    "check-msbuild": "ts-node -e "require(""./launcher/launcherCompiler"").checkMsbuildInPath(true)"",
    "check-node-version": "ts-node -e "require(""./utils/checkNodeVersion"").checkNodeRuntimeVersion()"",
    "node-sea": "npm run node-sea:build-blob && npm run node-sea:copy-node && npm run node-sea:unsign && npm run node-sea:inject-blob",
    "node-sea:build-blob": "node --experimental-sea-config sea-config.json",
    "node-sea:copy-node": "node -e "require('fs').copyFileSync(process.execPath, 'dist/test-app.exe')"",
    "node-sea:inject-blob": "postject dist/test-app.exe NODE_SEA_BLOB _blob\\sea-prep.blob --sentinel-fuse NODE_SEA_FUSE_fce680ab2cc467b6e072b8b5df1996b2",
    "node-sea:sign": "signtool sign /fd SHA256 dist/test-app.exe",
    "node-sea:unsign": "signtool remove /s dist/test-app.exe || echo Warning: signtool not found in "Path"",
    "pre-commit": "git diff HEAD --exit-code --stat launcher/* || npm run check-msbuild && npm run rebuild-launcher && git add resources/bin/test-app-launcher.exe",
    "prebuild": "npm run check-node-version",
    "prenode-sea:build-blob": "rimraf _blob && mkdir _blob",
    "prepare": "git config --get core.hookspath || husky install",
    "prewebpack": "rimraf build && rimraf dist",
    "rebuild-launcher": "msbuild launcher/launcher.csproj",
    "start": "ts-node src/index.ts",
    "type-check": "tsc --build tsconfig.json",
    "webpack": "webpack",
  },
  "version": "0.1.0",
}
`;

exports[`Test PackageJsonBuilder Test package.json with husky after javascript 1`] = `
{
  "main": "_build/index.js",
  "name": "test-app",
  "private": true,
  "scripts": {
    "build": "npm run webpack && npm run node-sea",
    "check-msbuild": "node -e "require(""./launcher/launcherCompiler"").checkMsbuildInPath(true)"",
    "check-node-version": "node -e "require(""./utils/checkNodeVersion"").checkNodeRuntimeVersion()"",
    "node-sea": "npm run node-sea:build-blob && npm run node-sea:copy-node && npm run node-sea:unsign && npm run node-sea:inject-blob",
    "node-sea:build-blob": "node --experimental-sea-config sea-config.json",
    "node-sea:copy-node": "node -e "require('fs').copyFileSync(process.execPath, 'dist/test-app.exe')"",
    "node-sea:inject-blob": "postject dist/test-app.exe NODE_SEA_BLOB _blob\\sea-prep.blob --sentinel-fuse NODE_SEA_FUSE_fce680ab2cc467b6e072b8b5df1996b2",
    "node-sea:sign": "signtool sign /fd SHA256 dist/test-app.exe",
    "node-sea:unsign": "signtool remove /s dist/test-app.exe || echo Warning: signtool not found in "Path"",
    "pre-commit": "git diff HEAD --exit-code --stat launcher/* || npm run check-msbuild && npm run rebuild-launcher && git add resources/bin/test-app-launcher.exe",
    "prebuild": "npm run check-node-version",
    "prenode-sea:build-blob": "rimraf _blob && mkdir _blob",
    "prepare": "git config --get core.hookspath || husky install",
    "prewebpack": "rimraf build && rimraf dist",
    "rebuild-launcher": "msbuild launcher/launcher.csproj",
    "start": "node src/index.js",
    "webpack": "webpack",
  },
  "version": "0.1.0",
}
`;

exports[`Test PackageJsonBuilder Test package.json with javascript 1`] = `
{
  "main": "_build/index.js",
  "name": "test-app",
  "private": true,
  "scripts": {
    "build": "npm run webpack && npm run node-sea",
    "check-msbuild": "node -e "require(""./launcher/launcherCompiler"").checkMsbuildInPath(true)"",
    "check-node-version": "node -e "require(""./utils/checkNodeVersion"").checkNodeRuntimeVersion()"",
    "node-sea": "npm run node-sea:build-blob && npm run node-sea:copy-node && npm run node-sea:unsign && npm run node-sea:inject-blob",
    "node-sea:build-blob": "node --experimental-sea-config sea-config.json",
    "node-sea:copy-node": "node -e "require('fs').copyFileSync(process.execPath, 'dist/test-app.exe')"",
    "node-sea:inject-blob": "postject dist/test-app.exe NODE_SEA_BLOB _blob\\sea-prep.blob --sentinel-fuse NODE_SEA_FUSE_fce680ab2cc467b6e072b8b5df1996b2",
    "node-sea:sign": "signtool sign /fd SHA256 dist/test-app.exe",
    "node-sea:unsign": "signtool remove /s dist/test-app.exe || echo Warning: signtool not found in "Path"",
    "prebuild": "npm run check-node-version",
    "prenode-sea:build-blob": "rimraf _blob && mkdir _blob",
    "prewebpack": "rimraf build && rimraf dist",
    "rebuild-launcher": "msbuild launcher/launcher.csproj",
    "start": "node src/index.js",
    "webpack": "webpack",
  },
  "version": "0.1.0",
}
`;

exports[`Test PackageJsonBuilder Test package.json with javascript after husky 1`] = `
{
  "main": "_build/index.js",
  "name": "test-app",
  "private": true,
  "scripts": {
    "build": "npm run webpack && npm run node-sea",
    "check-msbuild": "node -e "require(""./launcher/launcherCompiler"").checkMsbuildInPath(true)"",
    "check-node-version": "node -e "require(""./utils/checkNodeVersion"").checkNodeRuntimeVersion()"",
    "node-sea": "npm run node-sea:build-blob && npm run node-sea:copy-node && npm run node-sea:unsign && npm run node-sea:inject-blob",
    "node-sea:build-blob": "node --experimental-sea-config sea-config.json",
    "node-sea:copy-node": "node -e "require('fs').copyFileSync(process.execPath, 'dist/test-app.exe')"",
    "node-sea:inject-blob": "postject dist/test-app.exe NODE_SEA_BLOB _blob\\sea-prep.blob --sentinel-fuse NODE_SEA_FUSE_fce680ab2cc467b6e072b8b5df1996b2",
    "node-sea:sign": "signtool sign /fd SHA256 dist/test-app.exe",
    "node-sea:unsign": "signtool remove /s dist/test-app.exe || echo Warning: signtool not found in "Path"",
    "pre-commit": "git diff HEAD --exit-code --stat launcher/* || npm run check-msbuild && npm run rebuild-launcher && git add resources/bin/test-app-launcher.exe",
    "prebuild": "npm run check-node-version",
    "prenode-sea:build-blob": "rimraf _blob && mkdir _blob",
    "prepare": "git config --get core.hookspath || husky install",
    "prewebpack": "rimraf build && rimraf dist",
    "rebuild-launcher": "msbuild launcher/launcher.csproj",
    "start": "node src/index.js",
    "webpack": "webpack",
  },
  "version": "0.1.0",
}
`;
